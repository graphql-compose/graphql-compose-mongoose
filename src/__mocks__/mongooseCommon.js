/* @flow */
/* eslint-disable no-param-reassign, no-console */

import mongoose, { Schema } from 'mongoose';
import MongodbMemoryServer from 'mongodb-memory-server';

mongoose.Promise = Promise;

jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000;

const originalConnect = mongoose.connect;
mongoose.connect = async () => {
  const mongoServer = new MongodbMemoryServer();

  const mongoUri = await mongoServer.getConnectionString();

  originalConnect.bind(mongoose)(mongoUri, { useMongoClient: true });

  mongoose.connection.on('error', e => {
    if (e.message.code === 'ETIMEDOUT') {
      console.error(e);
    } else {
      throw e;
    }
  });

  mongoose.connection.once('open', () => {
    // console.log(`MongoDB successfully connected to ${mongoUri}`);
  });

  mongoose.connection.once('disconnected', () => {
    // console.log('MongoDB disconnected!');
    mongoServer.stop();
  });
};

export { mongoose, Schema };
