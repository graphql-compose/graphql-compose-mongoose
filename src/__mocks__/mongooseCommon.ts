/* eslint-disable no-param-reassign, no-console */

import mongoose from 'mongoose';
import MongoMemoryServer from 'mongodb-memory-server-core';
import net, { AddressInfo } from 'net';

const { Schema, Types } = mongoose;

mongoose.Promise = Promise;

export async function getPortFree() {
  return new Promise<number>((res) => {
    const srv = net.createServer();
    srv.listen(0, () => {
      const port = (srv.address() as AddressInfo).port;
      srv.close(() => res(port));
    });
  });
}

const originalConnect = mongoose.connect;
mongoose.createConnection = (async () => {
  const mongoServer = await MongoMemoryServer.create({
    instance: {
      port: await getPortFree(),
    },
  });
  const mongoUri = mongoServer.getUri();

  originalConnect.bind(mongoose)(mongoUri, {});

  mongoose.connection.on('error', (e) => {
    if (e.message.code === 'ETIMEDOUT') {
      console.error(e);
    } else {
      throw e;
    }
  });

  mongoose.connection.once('open', () => {
    // console.log(`MongoDB successfully connected to ${mongoUri}`);
  });

  mongoose.connection.once('disconnected', () => {
    // console.log('MongoDB disconnected!');
    mongoServer.stop();
  });
}) as any;

mongoose.connect = mongoose.createConnection as any;

export { mongoose, Schema, Types };
