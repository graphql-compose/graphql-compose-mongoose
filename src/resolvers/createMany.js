/* @flow */

import type { TypeComposer, Resolver } from 'graphql-compose';
import { graphql } from 'graphql-compose';
import type { MongooseModel } from 'mongoose';
import { recordHelperArgs } from './helpers';
import type { ExtendedResolveParams, GenResolverOpts } from './index';

async function createSingle(
  model: MongooseModel,
  tc: TypeComposer,
  recordData: any,
  resolveParams: ExtendedResolveParams
) {
  // eslint-disable-next-line new-cap
  let doc = new model(recordData);
  if (resolveParams.beforeRecordMutate) {
    doc = await resolveParams.beforeRecordMutate(doc, resolveParams);
    if (!doc) return null;
  }

  return doc.save();
}

export default function createMany(
  model: MongooseModel,
  tc: TypeComposer,
  opts?: GenResolverOpts
): Resolver {
  if (!model || !model.modelName || !model.schema) {
    throw new Error('First arg for Resolver createMany() should be instance of Mongoose Model.');
  }

  if (!tc || tc.constructor.name !== 'TypeComposer') {
    throw new Error('Second arg for Resolver createMany() should be instance of TypeComposer.');
  }

  const outputTypeName = `CreateMany${tc.getTypeName()}Payload`;
  const outputType = tc.constructor.schemaComposer.getOrCreateTC(outputTypeName, t => {
    t.addFields({
      recordIds: {
        type: '[MongoID]!',
        description: 'Created document ID',
      },
      records: {
        type: new graphql.GraphQLNonNull(tc.getTypePlural()),
        description: 'Created documents',
      },
      createCount: {
        type: 'Int!',
        description: 'Count of all documents created',
      },
    });
  });

  const resolver = new tc.constructor.schemaComposer.Resolver({
    name: 'createMany',
    kind: 'mutation',
    description: 'Creates Many documents with mongoose defaults, setters, hooks and validation',
    type: outputType,
    args: {
      records: {
        type: new graphql.GraphQLNonNull(
          graphql.GraphQLList(
            (recordHelperArgs(tc, {
              recordTypeName: `CreateMany${tc.getTypeName()}Input`,
              removeFields: ['id', '_id'],
              isRequired: true,
              ...(opts && opts.records),
            }).record: any).type
          )
        ),
      },
    },
    resolve: async (resolveParams: ExtendedResolveParams) => {
      const recordData = (resolveParams.args && resolveParams.args.records) || [];

      if (!Array.isArray(recordData) || recordData.length === 0) {
        throw new Error(
          `${tc.getTypeName()}.createMany resolver requires args.records to be an Array and must contain at least one record`
        );
      }

      for (const record of recordData) {
        if (!(typeof record === 'object') || Object.keys(record).length === 0) {
          throw new Error(
            `${tc.getTypeName()}.createMany resolver requires args.records to contain non-empty records, with at least one value`
          );
        }
      }

      const recordPromises = [];
      // concurrently create docs
      for (const record of recordData) {
        recordPromises.push(createSingle(model, tc, record, resolveParams));
      }

      const results = await Promise.all(recordPromises);
      const returnObj = {
        records: [],
        recordIds: [],
        createCount: 0,
      };

      for (const doc of results) {
        if (doc) {
          returnObj.createCount += 1;
          returnObj.records.push(doc);
          returnObj.recordIds.push(doc._id);
        }
      }

      return returnObj;
    },
  });

  return resolver;
}
