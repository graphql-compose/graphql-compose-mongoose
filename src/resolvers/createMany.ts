import { ObjectTypeComposer, Resolver } from 'graphql-compose';
import type { Model, Document } from 'mongoose';
import { recordHelperArgs, RecordHelperArgsOpts } from './helpers';
import { addErrorCatcherField } from './helpers/errorCatcher';
import { validateManyAndThrow } from './helpers/validate';
import { payloadRecordIds, PayloadRecordIdsHelperOpts } from './helpers/payloadRecordId';

export interface CreateManyResolverOpts {
  /** If you want to generate different resolvers you may avoid Type name collision by adding a suffix to type names */
  suffix?: string;
  /** Customize input-type for `records` argument. */
  records?: RecordHelperArgsOpts;
  /** Customize payload.recordIds field. If false, then this field will be removed. */
  recordIds?: PayloadRecordIdsHelperOpts | false;
}

type TArgs = {
  records: any[];
};

export function createMany<TSource = any, TContext = any, TDoc extends Document = any>(
  model: Model<TDoc>,
  tc: ObjectTypeComposer<TDoc, TContext>,
  opts?: CreateManyResolverOpts
): Resolver<TSource, TContext, TArgs, TDoc> {
  if (!model || !model.modelName || !model.schema) {
    throw new Error('First arg for Resolver createMany() should be instance of Mongoose Model.');
  }

  if (!tc || tc.constructor.name !== 'ObjectTypeComposer') {
    throw new Error(
      'Second arg for Resolver createMany() should be instance of ObjectTypeComposer.'
    );
  }

  const tree = model.schema.obj;
  const requiredFields = [];
  for (const field in tree) {
    if (tree.hasOwnProperty(field)) {
      const fieldOptions = tree[field];
      if (fieldOptions.required && typeof fieldOptions.required !== 'function') {
        requiredFields.push(field);
      }
    }
  }

  const outputTypeName = `CreateMany${tc.getTypeName()}${opts?.suffix || ''}Payload`;
  const outputType = tc.schemaComposer.getOrCreateOTC(outputTypeName, (t) => {
    t.setFields({
      ...payloadRecordIds(tc, opts?.recordIds),
      records: {
        type: tc.NonNull.List,
        description: 'Created documents',
      },
      createdCount: {
        type: 'Int!',
        description: 'Number of created documents',
        resolve: (s: any) => s.createdCount || 0,
      },
    });
  });

  const resolver = tc.schemaComposer.createResolver<TSource, TArgs>({
    name: 'createMany',
    kind: 'mutation',
    description: 'Creates Many documents with mongoose defaults, setters, hooks and validation',
    type: outputType,
    args: {
      records: {
        type: (recordHelperArgs(tc, {
          prefix: 'CreateMany',
          suffix: `${opts?.suffix || ''}Input`,
          removeFields: ['id', '_id'],
          isRequired: true,
          requiredFields,
          ...opts?.records,
        }) as any).record.type.List.NonNull,
      },
    },
    resolve: async (resolveParams) => {
      const recordData = resolveParams?.args?.records;

      if (!Array.isArray(recordData) || recordData.length === 0) {
        throw new Error(
          `${tc.getTypeName()}.createMany resolver requires args.records to be an Array and must contain at least one record`
        );
      }

      for (const record of recordData) {
        if (!(typeof record === 'object') || Object.keys(record).length === 0) {
          throw new Error(
            `${tc.getTypeName()}.createMany resolver requires args.records to contain non-empty records, with at least one value`
          );
        }
      }

      const docs = [] as TDoc[];
      for (const record of recordData) {
        // eslint-disable-next-line new-cap
        let doc = new model(record);
        if (resolveParams.beforeRecordMutate) {
          doc = await resolveParams.beforeRecordMutate(doc, resolveParams);
        }
        docs.push(doc);
      }

      await validateManyAndThrow(docs);
      await model.create(docs as any, { validateBeforeSave: false });

      return {
        records: docs,
        createdCount: docs.length,
      };
    },
  });

  // Add `error` field to payload which can catch resolver Error
  // and return it in mutation payload
  addErrorCatcherField(resolver);

  return resolver;
}
