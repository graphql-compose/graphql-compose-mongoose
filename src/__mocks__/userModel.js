/* @flow */

import { mongoose, Schema } from './mongooseCommon';
import ContactsSchema from './contactsSchema';
import enumEmployment from './enumEmployment';
import LanguageSchema from './languageSchema';

const UserSchema = new Schema(
  {
    subDoc: {
      field1: String,
      field2: {
        field21: String,
        field22: String,
      },
    },

    user: {
      type: Schema.Types.ObjectId,
      ref: 'UserModel',
    },

    users: {
      type: [Schema.Types.ObjectId],
      ref: 'UserModel',
    },

    name: {
      type: String,
      description: 'Person name',
    },

    age: {
      type: Number,
      description: 'Full years',
    },

    gender: {
      type: String,
      enum: ['male', 'female', 'ladyboy'],
    },

    skills: {
      type: [String],
      default: [],
      description: 'List of skills',
    },

    employment: {
      type: [
        {
          type: String,
          enum: Object.keys(enumEmployment),
        },
      ],
      description: 'List of desired employment types',
      index: true,
    },

    relocation: {
      type: Boolean,
      description: 'Does candidate relocate to another region',
    },

    contacts: {
      type: ContactsSchema,
      default: {},
      description: 'Contacts of person (phone, skype, mail and etc)',
    },

    languages: {
      type: [LanguageSchema],
      default: [],
      description: 'Knowledge of languages',
    },

    __secretField: {
      type: String,
    },

    someDynamic: {
      type: Schema.Types.Mixed,
      description: "Some mixed value, that served with @taion's `graphql-type-json`",
    },

    // createdAt, created via option `timastamp: true` (see bottom)
    // updatedAt, created via option `timastamp: true` (see bottom)
  },
  {
    timestamps: true, // add createdAt, updatedAt fields
    toJSON: { getters: true },
    toObject: { virtuals: true },
  }
);

UserSchema.set('autoIndex', false);
UserSchema.index({ name: 1, age: -1 });

UserSchema.virtual('nameVirtual').get(function () { // eslint-disable-line
  return `VirtualFieldValue${this._id}`;
});

const UserModel = mongoose.model('User', UserSchema);

export { UserSchema, UserModel };
