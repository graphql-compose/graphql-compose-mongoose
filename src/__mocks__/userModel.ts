import { mongoose, Schema } from './mongooseCommon';
import ContactsSchema from './contactsSchema';
import enumEmployment from './enumEmployment';
import LanguageSchema from './languageSchema';
import { Document } from 'mongoose';

const UserSchema = new Schema(
  {
    subDoc: {
      field1: String,
      field2: {
        field21: String,
        field22: String,
      },
    },

    user: {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },

    users: {
      type: [Schema.Types.ObjectId],
      ref: 'User',
    },

    n: {
      type: String,
      required: true,
      description: 'Person name',
      alias: 'name',
    },

    age: {
      type: Number,
      description: 'Full years',
      required() {
        // in graphql this field should be Nullable
        return (this as any).name === 'Something special';
      },
    },

    gender: {
      type: String,
      enum: ['male', 'female', 'ladyboy'],
    },

    skills: {
      type: [String],
      default: [],
      description: 'List of skills',
    },

    employment: {
      type: [
        {
          type: String,
          enum: Object.keys(enumEmployment),
        },
      ],
      description: 'List of desired employment types',
      index: true,
    },

    relocation: {
      type: Boolean,
      description: 'Does candidate relocate to another region',
    },

    contacts: {
      type: ContactsSchema,
      default: {},
      description: 'Contacts of person (phone, skype, mail and etc)',
    },

    languages: {
      type: [LanguageSchema],
      default: [],
      description: 'Knowledge of languages',
    },

    __secretField: {
      type: String,
    },

    someDynamic: {
      type: Schema.Types.Mixed,
      description: "Some mixed value, that served with @taion's `graphql-type-json`",
    },

    periods: [{ from: Number, to: Number }],

    someDeep: {
      periods: [{ from: Number, to: Number }],
    },

    salary: {
      type: Schema.Types.Decimal128,
    },

    mapField: {
      type: Map,
      of: String,
    },

    mapFieldDeep: {
      subField: {
        type: Map,
        of: String,
      },
    },

    billingAddress: {
      street: { type: String, index: true },
      state: { type: String, index: true, enum: ['FL', 'MA', 'NY'] },
      country: { type: String, index: true },
    },

    // for error payloads tests
    valid: {
      type: String,
      required: false,
      validate: [
        () => {
          return false;
        },
        'this is a validate message',
      ],
    },

    // createdAt, created via option `timestamp: true` (see bottom)
    // updatedAt, created via option `timestamp: true` (see bottom)
  },
  {
    timestamps: true, // add createdAt, updatedAt fields
    toJSON: { getters: true },
    toObject: { virtuals: true },
  }
);

UserSchema.set('autoIndex', false);
UserSchema.index({ name: 1, age: -1 });

// eslint-disable-next-line
UserSchema.virtual('nameVirtual').get(function (this: any) {
  return `VirtualFieldValue${this._id}`;
});

export interface IUser extends Document {
  _id: any;
  name?: string;
  age?: number;
  gender?: string;
  someDynamic?: any;
  skills?: string[];
  relocation?: boolean;
  contacts: {
    email: string;
    skype?: string;
  };
}

const UserModel = mongoose.model<IUser>('User', UserSchema);

export { UserSchema, UserModel };
