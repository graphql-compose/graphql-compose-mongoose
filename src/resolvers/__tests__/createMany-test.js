/* @flow */
/* eslint-disable no-param-reassign,func-names */

import { Resolver, schemaComposer, TypeComposer } from 'graphql-compose';
import { GraphQLInt, GraphQLList, GraphQLNonNull } from 'graphql-compose/lib/graphql';
import { mongoose } from '../../__mocks__/mongooseCommon';
import { UserModel } from '../../__mocks__/userModel';
import { convertModelToGraphQL } from '../../fieldsConverter';
import GraphQLMongoID from '../../types/mongoid';
import createMany from '../createMany';

beforeAll(() => UserModel.base.connect());
afterAll(() => UserModel.base.disconnect());

describe('createMany() ->', () => {
  let UserTC;

  beforeEach(() => {
    schemaComposer.clear();
    UserTC = convertModelToGraphQL(UserModel, 'User', schemaComposer);
    UserTC.setRecordIdFn(source => (source ? `${source._id}` : ''));
  });

  beforeEach(async () => {
    await UserModel.remove({});
  });

  it('should return Resolver object', () => {
    const resolver = createMany(UserModel, UserTC);
    expect(resolver).toBeInstanceOf(Resolver);
  });

  describe('Resolver.args', () => {
    it('should have required `records` arg', () => {
      const resolver = createMany(UserModel, UserTC);
      const argConfig: any = resolver.getArgConfig('records');
      expect(argConfig.type).toBeInstanceOf(GraphQLNonNull);
      expect(argConfig.type.toString()).toBe('[CreateManyUserInput!]!');
    });
    it('should have `records` arg as Plural', () => {
      const resolver = createMany(UserModel, UserTC);
      const argConfig: any = resolver.getArgConfig('records');
      expect(argConfig.type.ofType).toBeInstanceOf(GraphQLList);
      expect(argConfig.type.ofType.toString()).toBe('[CreateManyUserInput!]');
    });
    it('should have `records` arg internal item as required', () => {
      const resolver = createMany(UserModel, UserTC);
      const argConfig: any = resolver.getArgConfig('records');
      expect(argConfig.type).toBeInstanceOf(GraphQLNonNull);
      expect(argConfig.type.ofType.ofType.toString()).toBe('CreateManyUserInput!');
    });
  });

  describe('Resolver.resolve():Promise', () => {
    it('should be promise', () => {
      const result = createMany(UserModel, UserTC).resolve({});
      expect(result).toBeInstanceOf(Promise);
      result.catch(() => 'catch error if appear, hide it from mocha');
    });

    it('should rejected with Error if args.records is empty', async () => {
      const result = createMany(UserModel, UserTC).resolve({ args: {} });
      await expect(result).rejects.toMatchSnapshot();
    });

    it('should rejected with Error if args.records is not array', async () => {
      const result = createMany(UserModel, UserTC).resolve({ args: { records: {} } });
      await expect(result).rejects.toMatchSnapshot();
    });

    it('should rejected with Error if args.records is empty array', async () => {
      const result = createMany(UserModel, UserTC).resolve({
        args: { records: [] },
      });
      await expect(result).rejects.toMatchSnapshot();
    });

    it('should rejected with Error if args.records is array with empty items', async () => {
      const result = createMany(UserModel, UserTC).resolve({
        args: { records: [{ name: 'fails' }, {}] },
      });
      await expect(result).rejects.toMatchSnapshot();
    });

    it('should return payload.recordIds', async () => {
      const result = await createMany(UserModel, UserTC).resolve({
        args: {
          records: [{ name: 'newName' }],
        },
      });
      expect(result.recordIds).toBeTruthy();
    });

    it('should create documents with args.records and match createCount', async () => {
      const result = await createMany(UserModel, UserTC).resolve({
        args: {
          records: [{ name: 'newName0' }, { name: 'newName1' }],
        },
      });
      expect(result.createCount).toBe(2);
      expect(result.records[0].name).toBe('newName0');
      expect(result.records[1].name).toBe('newName1');
    });

    it('should save documents to database', async () => {
      const checkedName = 'nameForMongoDB';
      const res = await createMany(UserModel, UserTC).resolve({
        args: {
          records: [{ name: checkedName }, { name: checkedName }],
        },
      });

      const docs = await UserModel.collection.find({ _id: { $in: res.recordIds } }).toArray();
      expect(docs.length).toBe(2);
      expect(docs[0].name).toBe(checkedName);
      expect(docs[1].name).toBe(checkedName);
    });

    it('should return payload.records', async () => {
      const result = await createMany(UserModel, UserTC).resolve({
        args: {
          records: [{ name: 'NewUser' }],
        },
      });
      expect(result.records[0]._id).toBe(result.recordIds[0]);
    });

    it('should return mongoose documents', async () => {
      const result = await createMany(UserModel, UserTC).resolve({
        args: { records: [{ name: 'NewUser' }] },
      });
      expect(result.records[0]).toBeInstanceOf(UserModel);
    });

    it('should call `beforeRecordMutate` method with each created `record` and `resolveParams` as args', async () => {
      const result = await createMany(UserModel, UserTC).resolve({
        args: { records: [{ name: 'NewUser0' }, { name: 'NewUser1' }] },
        context: { ip: '1.1.1.1' },
        beforeRecordMutate: (record, rp) => {
          record.name = 'OverridedName';
          record.someDynamic = rp.context.ip;
          return record;
        },
      });
      expect(result.records[0]).toBeInstanceOf(UserModel);
      expect(result.records[1]).toBeInstanceOf(UserModel);
      expect(result.records[0].name).toBe('OverridedName');
      expect(result.records[1].name).toBe('OverridedName');
      expect(result.records[0].someDynamic).toBe('1.1.1.1');
      expect(result.records[1].someDynamic).toBe('1.1.1.1');
    });

    it('should execute hooks on save', async () => {
      schemaComposer.clear();
      const ClonedUserSchema = UserModel.schema.clone();

      ClonedUserSchema.pre('save', function(next) {
        this.name = 'ChangedAgain';
        this.age = 18;
        return next();
      });

      const ClonedUserModel = mongoose.model('UserClone', ClonedUserSchema);

      const ClonedUserTC = convertModelToGraphQL(ClonedUserModel, 'UserClone', schemaComposer);
      ClonedUserTC.setRecordIdFn(source => (source ? `${source._id}` : ''));

      const result = await createMany(ClonedUserModel, ClonedUserTC).resolve({
        args: { records: [{ name: 'NewUser0' }, { name: 'NewUser1' }] },
        context: { ip: '1.1.1.1' },
        beforeRecordMutate: (record, rp) => {
          record.name = 'OverridedName';
          record.someDynamic = rp.context.ip;
          return record;
        },
      });
      expect(result.records[0]).toBeInstanceOf(ClonedUserModel);
      expect(result.records[1]).toBeInstanceOf(ClonedUserModel);
      expect(result.records[0].age).toBe(18);
      expect(result.records[1].age).toBe(18);
      expect(result.records[0].name).toBe('ChangedAgain');
      expect(result.records[1].name).toBe('ChangedAgain');
      expect(result.records[0].someDynamic).toBe('1.1.1.1');
      expect(result.records[1].someDynamic).toBe('1.1.1.1');
    });
  });

  describe('Resolver.getType()', () => {
    it('should have correct output type name', () => {
      const outputType: any = createMany(UserModel, UserTC).getType();
      expect(outputType.name).toBe(`CreateMany${UserTC.getTypeName()}Payload`);
    });

    it('should have recordIds field, NonNull List', () => {
      const outputType: any = createMany(UserModel, UserTC).getType();
      const recordIdField = new TypeComposer(outputType).getFieldConfig('recordIds');
      expect(recordIdField.type).toEqual(new GraphQLNonNull(GraphQLList(GraphQLMongoID)));
    });

    it('should have records field, NonNull List', () => {
      const outputType: any = createMany(UserModel, UserTC).getType();
      const recordField = new TypeComposer(outputType).getFieldConfig('records');
      expect(recordField.type).toEqual(new GraphQLNonNull(GraphQLList(UserTC.getType())));
    });

    it('should have createCount field, Int', () => {
      const outputType: any = createMany(UserModel, UserTC).getType();
      const recordField = new TypeComposer(outputType).getFieldConfig('createCount');
      expect(recordField.type).toEqual(new GraphQLNonNull(GraphQLInt));
    });

    it('should reuse existed outputType', () => {
      const outputTypeName = `CreateMany${UserTC.getTypeName()}Payload`;
      const existedType = TypeComposer.create(outputTypeName);
      schemaComposer.set(outputTypeName, existedType);
      const outputType = createMany(UserModel, UserTC).getType();
      expect(outputType).toBe(existedType.getType());
    });
  });
});
