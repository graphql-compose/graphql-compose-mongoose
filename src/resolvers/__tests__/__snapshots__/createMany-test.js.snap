// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMany() -> Resolver.resolve():Promise should rejected with Error if args.records is array with empty items 1`] = `[Error: User.createMany resolver requires args.records to contain non-empty records, with at least one value]`;

exports[`createMany() -> Resolver.resolve():Promise should rejected with Error if args.records is empty 1`] = `[Error: User.createMany resolver requires args.records to be an Array and must contain at least one record]`;

exports[`createMany() -> Resolver.resolve():Promise should rejected with Error if args.records is empty array 1`] = `[Error: User.createMany resolver requires args.records to be an Array and must contain at least one record]`;

exports[`createMany() -> Resolver.resolve():Promise should rejected with Error if args.records is not array 1`] = `[Error: User.createMany resolver requires args.records to be an Array and must contain at least one record]`;
